ssm/instance/fetch.go

```diff
@@ -1,54 +1,89 @@
 package instance
 
 import (
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ssm"
 	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
+	"github.com/disneystreaming/go-ssmhelpers/util/batch"
 )
 
 // getSSMInstances takes an SSM session and an *ssm.DescribeInstanceInformationInput object to get information about a given SSM-managed EC2 instance.
 func getSSMInstances(session ssmiface.SSMAPI, input *ssm.DescribeInstanceInformationInput) (output *ssm.DescribeInstanceInformationOutput, err error) {
 	return session.DescribeInstanceInformation(input)
 }
 
-// GetAllSSMInstances takes an SSM session and an *ssm.DescribeInstanceInformationInput object, and calls getSSMInstances() to query the SSM API for
-// information about a given SSM-managed EC2 instance. It also filters out any instances returned that are unresponsive to ping or not running Linux.
-// Filtering can only be done based on either instance information OR instance tags for any given instance of a query. As such, filtering of instances
-// by things like bamazon and env tags must be done after retrieving the instance information.
-func GetAllSSMInstances(session ssmiface.SSMAPI, input *ssm.DescribeInstanceInformationInput, checkLatestAgent bool) (output []*ssm.InstanceInformation, err error) {
-	// First call to set up NextToken
+// getSSMInstances takes an SSM session and an *ssm.DescribeInstanceInformationInput object to get information about a given SSM-managed EC2 instance.
+func getSSMInstancesPages(session ssmiface.SSMAPI, input *ssm.DescribeInstanceInformationInput) (output []*ssm.DescribeInstanceInformationOutput, err error) {
 	results, err := getSSMInstances(session, input)
+	if err != nil {
+		return nil, err
+	}
 
-	var allResults []*ssm.InstanceInformation
-	for _, v := range results.InstanceInformationList {
-		// Remove any instances that are unresponsive to ping or not running Linux
-		if checkLatestAgent {
-			if *v.PingStatus == "Online" && *v.PlatformType == "Linux" && *v.IsLatestVersion {
-				allResults = append(allResults, v)
-			}
-		} else {
-			if *v.PingStatus == "Online" && *v.PlatformType == "Linux" {
-				allResults = append(allResults, v)
-			}
+	output = append(output, results)
+
+	for results.NextToken != nil && *results.NextToken != "" && err == nil {
+		input.SetNextToken(*results.NextToken)
+		results, err = getSSMInstances(session, input)
+		if err != nil {
+			return output, err
+		}
+		output = append(output, results)
+	}
+
+	return output, err
+}
+
+// GetAllSSMInstances queries the SSM API for information about SSM-managed instances, and filters out any instances that are unresponsive to ping or not running Linux.
+// It can also check to determine if instances are running the latest SSM agent version, which is a prerequisite for connecting via start-session functionality.
```

Comment:
Why is latest version a prereq for connecting with start-session? According to the docs the minimum version is 2.3.68.0 https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-prerequisites.html